### PR TITLE
Fix Command Line Parsing Bug

### PR DESCRIPTION
**Pull request by bmathwig**
The command line parsing logic will fail in the following scenario:
``` bash
php -d setting=true -d directory=/usr/lib -R 'sleep(1024);'
```

Two problems occur when inspecting this process:
* The loop ends early, detecting `-d setting=true` to be the start of environment variables
* The code `File.basename()` will truncate the string to `lib -R 'sleep(100000);'`

This pull request attempts to fix both of them. The fixes include:
* Adding a call to sysctl using the KERN_PROCARGS2 argument to get the argc value
* Looping until argc value to gather all command line arguments into the cmdline variable
* Adding a unit test to support the above edge case

### CODE REVIEW

**Comment by bmathwig** [comment]
Working on a solid unit test case to solve this issue. It looks to be much more complicated that just removing the File.basename() call. A good example to use is:

`> php -d something=hello -d library=/usr/lib -R 'sleep(100000);'`

This results in a cmdline variable equal to 'php', which is incorrect.

**Comment by djberg96** [comment]
@bmathwig I think we need an actual specification if we can find one so we can test against that.

**Comment by bmathwig** [comment]
Take a look at the new changes I made. I've updated the code to properly loop the argument list using 'argc', splitting the arguments and environment variables appropriately without using '=' as a terminator.

**Comment by djberg96** [comment]
@bmathwig Ooh, this is looking good. I don't know about using "php" as a test, though, since it's not guaranteed to be installed (is it?). Is there some standard OSX system process we could use instead?

File: lib/darwin/sys/proctable.rb [hunk 1]
```diff
@@ -273,6 +274,18 @@ def self.get_cmd_args_and_env(pid, struct)
...
+      mib.write_array_of_int([CTL_KERN, KERN_PROCARGS2, pid])
+      return if sysctl(mib, 3, nil, len, nil, 0) < 0
+
+      buf = FFI::MemoryPointer.new(:char, len.read_ulong)
+      return if sysctl(mib, 3, buf, len, nil, 0) < 0
+
+      # The argc value is located in the first byte of buf
+      argc = buf.read_bytes(1).ord
+      buf.free
+
```
Comment (djberg96) [inline lib/darwin/sys/proctable.rb]: Since this is a one time buffer, I think the block form would be appropriate here.

    FFI::MemoryPointer.new(:char, len.read_ulong) do |buf|
        ...
    end

File: test/test_sys_proctable_darwin.rb [hunk 2]
```diff
@@ -23,11 +23,16 @@ def self.startup
       exec('env', '-i', 'A=B', 'Z=', 'sleep', '60')
     end
 
+    @@pid2 = fork do
+      exec('php', '-d setting=true', '-d directory=/usr/bin', '-R', '\'sleep(120);\'')
+    end
+
```
Comment (djberg96) [inline test/test_sys_proctable_darwin.rb]: As I mentioned, I think we should try to find a system process that we can use instead. If we do use an external process, I tend to lean towards `Process.spawn` since it gives you more control options.

**Comment by bmathwig** [comment]
PHP is shipped with OSX, so I think we should be OK assuming it is there.

**Comment by djberg96** [comment]
Alright, I'll merge this and tweak it later. :)

Thank you very much for this!
